Specfem-Core-Framework-Logger-Impl

lsawade opened the pull request:
## Description

Implements the logger struct and moves some things into the discussed namespaces.

- Added the logger and the logger options structs (b14e64c4)
- Deprecated the context - context guard combination in favor of a single context struct (5d26ac2b)
- Forgot to add the header file (5b1316f0)
- Add new static struct specfem::MPI that holds rank, size, and comm information. (bfe6c98b)
- Renamed simulation to program and moved context namespace. (27e02149)

## Issue Number

Updates #1149

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

icui committed:
Remove write_adjacency_map parameter.

icui committed:
Snakefile for HomogeneousElasticDomain test.

icui committed:
Update tests/unit-tests/displacement_tests/Newmark/Snakefile

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

icui committed:
Update tests/unit-tests/displacement_tests/Newmark/Snakefile

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Rohit-Kakodkar committed:
Merge pull request #1318 from PrincetonUniversity/1178-meshpar

1178 Part I - Remove write_adjacency_map flag

icui committed:
Update snakemake.

icui committed:
Check individual trace files.

Rohit-Kakodkar committed:
Issue 1281 - Compute material properties using MESHFFEM3D mesh

- [x] Adds a dimension tag to materials
- [x] Adds element tags within the mesher
- [x] Adds constructor to element types
- [x] Adds constructor to assembly properties

icui committed:
Update trace list snakefile

icui committed:
Move snakefile

icui committed:
Check trace list for displacement test rule.

Rohit-Kakodkar committed:
Fixed device access

Rohit-Kakodkar committed:
Issue 1334 - Construct assembly using MESHFEM3D mesh

- [x] Adds new constructor for receivers
- [x] MESHFEM3D mesher now writes number of NGLL points
- [x] Added element grid to mesh struct
- [x] Added test to check element grid

icui committed:
Re-organize specfem smk

icui committed:
Update rule.

icui committed:
Remove unnecessary binary linking.

Rohit-Kakodkar committed:
review comments

Rohit-Kakodkar committed:
review comments

icui committed:
Update snakemake rules to avoid rerunning.

int-ptr-ptr committed:
additional accessors necessary for dg2d kernel

Copilot committed:
Initial plan

Copilot committed:
Add force install functionality for ADIOS2 cmake

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Update CMakePresets.json to include ADIOS2 force install option in release-io preset

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Fix code review issues: improve cmake policy logic and remove unreachable code

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Add safety fallback else clause for consistency with HDF5 implementation

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Fix cmake policy logic: only set CMP0169 when using FetchContent_MakeAvailable

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Add target existence checks before creating alias targets for robustness

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Improve comments to clarify CMake policy logic and version-specific behavior

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

Copilot committed:
Add force install functionality for Boost library

Co-authored-by: Rohit-Kakodkar <40695390+Rohit-Kakodkar@users.noreply.github.com>

int-ptr-ptr committed:
removed `shmem_size` from point accessors

Rohit-Kakodkar committed:
Merge pull request #1335 from PrincetonUniversity/issue-1334

Issue 1334 - Construct assembly using MESHFEM3D mesh

icui committed:
Fix snakemake rules.

int-ptr-ptr committed:
isolated kernel from dg2d-kernelimpl

int-ptr-ptr committed:
src from kernelimpl, remove edge_to_interface

int-ptr-ptr committed:
add nc displacement test

icui committed:
Merge pull request #1323 from PrincetonUniversity/1178-snake

Snakefile for HomogeneousElasticDomain test.

icui committed:
Merge branch 'devel' into issue-1178

int-ptr-ptr committed:
replaced get_transfer_function method with reference

icui committed:
Link tests.yaml

icui committed:
Generate database from specfempp

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

int-ptr-ptr committed:
gpu not compiling again -- points to ncmarked_conforming_grid_flat

icui committed:
Fix linking and rules.

icui committed:
Update clean rule.

Rohit-Kakodkar committed:
Merge branch 'devel' into copilot/fix-1919273-506020639-6165e7bc-b1ab-4b84-a5df-aafa63a354dd

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1281

int-ptr-ptr committed:
Merge branch 'devel' into nci-accessor

icui committed:
Use static snake files for tests.

icui committed:
Clean up.

int-ptr-ptr committed:
modified cmakelists

lsawade committed:
Updated badges

icui committed:
Merge branch 'devel' into 1178-2d

icui committed:
Fix benchmarks params

Rohit-Kakodkar committed:
Fix bug in coupled interfaces

- [x] Update index function in edge/intersection iterator
- [x] Update `for_all` execution function

int-ptr-ptr committed:
removed scratch constructor from point

int-ptr-ptr committed:
removed simd

int-ptr-ptr committed:
removed simd from test -- compiles now

Rohit-Kakodkar committed:
Add `KOKKOS_PATH` flag to CMake

This PR adds support for using a local installation of Kokkos when building SPECFEM++

icui committed:
Update HomogeneousElasticDomainStaceyBC

Rohit-Kakodkar committed:
compiler fix

icui committed:
Add snakefiles

icui committed:
Add snakefile to tests.

icui committed:
Add snakefile to tests.

Rohit-Kakodkar committed:
Adds cmake unity build flag

icui committed:
Fix test stations.

icui committed:
Merge branch 'devel' into 1178-db

icui committed:
Merge branch 'devel' into issue-1178

icui committed:
Update electromagnetic test snakefile.

int-ptr-ptr committed:
compatibility with nci-accessor changes

int-ptr-ptr committed:
review changes

lsawade committed:
Implemented compute_source_array for tensor, as well as associated test.

lsawade committed:
Updated the Jenkins workflow to install in separate directories.

icui committed:
Merge branch 'issue-1178' into 1178-db

int-ptr-ptr committed:
polarized transfer algorithm, removed ambiguated transfer_function ref

Rohit-Kakodkar committed:
Added documentation in getting started

Rohit-Kakodkar committed:
Added support for pre-installed kokkos library

Rohit-Kakodkar committed:
Merge branch 'issue-1328' into issue-1331

icui committed:
Update adjacency reader and database.

icui committed:
Update tests/unit-tests/displacement_tests/Newmark/README.md

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

icui committed:
Merge pull request #1346 from PrincetonUniversity/1178-db

1178 - Generate specfempp database from snakemake and make Snakefile static

Rohit-Kakodkar committed:
Merge pull request #1352 from PrincetonUniversity/issue-1331

Add `KOKKOS_PATH` flag to CMake

Rohit-Kakodkar committed:
Refactors all macros into single library

- [x] Adds a macros library in core
- [x] Moves all macros from enumerations, macros to this library
- [x] Removes `__future__`

Rohit-Kakodkar committed:
Removed macros file

icui committed:
Snakefile for test data

icui committed:
Update main test snakefile

icui committed:
Merge branch '1178-data' into 1178-2d

icui committed:
Update stacey test.

icui committed:
Update finite difference workflow.

icui committed:
Use trace linking for fd.

icui committed:
Finalize 2d tests.

icui committed:
Update snakemake linking.

icui committed:
Link individual files for fd.

Rohit-Kakodkar committed:
Add cmake flag to nc grid

Rohit-Kakodkar committed:
Merge branch 'fix-cmake-ncgrid' into issue-1281

Rohit-Kakodkar committed:
Fixed macros

Rohit-Kakodkar committed:
Merge branch 'fix-cmake-ncgrid' into macros-overhaul

lsawade committed:
Merge pull request #1341 from PrincetonUniversity/copilot/fix-1919273-506020639-6165e7bc-b1ab-4b84-a5df-aafa63a354dd

Add force install options for ADIOS2 and Boost dependencies

lsawade committed:
Merge pull request #1348 from PrincetonUniversity/badge-fix

Updated README.md badges

lsawade committed:
Merge pull request #1349 from PrincetonUniversity/benchmarks-fix

Fix snakemake params in benchmarks folder

Rohit-Kakodkar committed:
Merge pull request #1360 from PrincetonUniversity/fix-cmake-ncgrid

Add cmake flag to nc grid

Rohit-Kakodkar committed:
review comments

Rohit-Kakodkar committed:
Added missing file

Rohit-Kakodkar committed:
Minor fix

Rohit-Kakodkar committed:
Make Meshfem3D database the default database for 3D

- [x] Updates parameter parser
- [x] Updates the execute routine for 3d
- [x] Update 3D tests and databases

icui committed:
Add finite difference workflow.

icui committed:
Re-organize.

icui committed:
Update trace output.

Rohit-Kakodkar committed:
Merge pull request #1329 from PrincetonUniversity/issue-1281

Issue 1281 - Compute material properties using MESHFFEM3D mesh

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1350

int-ptr-ptr committed:
Merge remote-tracking branch 'public_repo/issue-1350' into nci-accessor

int-ptr-ptr committed:
ncmarked_conforming_grid_curved: rel src

revert  ncmarked_conforming_grid_curved CMakeLists.txt sources to relative (to match ncmarked_conforming_grid_flat CmakeLists)

int-ptr-ptr committed:
Merge branch 'nci-accessor' into nci-kernel

int-ptr-ptr committed:
resolved post-merge compilation issues

int-ptr-ptr committed:
fixed newmark displacement test config

int-ptr-ptr committed:
bugfixes post-merge

Rohit-Kakodkar committed:
Merge pull request #1351 from PrincetonUniversity/issue-1350

Fix bug in coupled interfaces

Rohit-Kakodkar committed:
Merge pull request #1339 from int-ptr-ptr/nci-accessor

NCI accessors: chunk_edge (normal, intersection_factor) and point (transfer_function)

Rohit-Kakodkar committed:
Merge pull request #1343 from int-ptr-ptr/nci-kernel

Nonconforming Kernel

icui committed:
Update snakemake required version.

Rohit-Kakodkar committed:
Update core/specfem/macros/CMakeLists.txt

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

lsawade committed:
Updated the nightly benchmarks to include git info

Rohit-Kakodkar committed:
Merge pull request #1354 from PrincetonUniversity/issue-1219

Issue 1219 - Implemented compute_source_array for tensor, as well as associated test.

Rohit-Kakodkar committed:
Merge pull request #1355 from PrincetonUniversity/jenkins-install-fix

Updated the Jenkins workflow to install in separate directories.

Rohit-Kakodkar committed:
Merge pull request #1356 from PrincetonUniversity/issue-1328

Adds a Unity build batch size flag

Rohit-Kakodkar committed:
Coupled interface bugfixes

This PR fixes bug in coupled interfaces where the coupling kernel would only take weight and factor for the first edge. The PR also generalizes the Edge & Interface iterators such that 3D interfaces can be incorporated.

- [x] Updates to mesh entities in 2D
- [x] Updates to connection mapping in 2D
- [x] Updates edge types

Rohit-Kakodkar committed:
Merge branch 'devel' into macros-overhaul

Rohit-Kakodkar committed:
Fixed minor bug

Rohit-Kakodkar committed:
Added missing database

Rohit-Kakodkar committed:
Merge branch 'macros-overhaul' into issue-1062

Rohit-Kakodkar committed:
Added runtime errors

icui committed:
Add intrinsic spin and curl

icui committed:
Update test data.

Rohit-Kakodkar committed:
Merge pull request #1357 from PrincetonUniversity/macros-overhaul

Refactors all macros into single library

Rohit-Kakodkar committed:
Merge pull request #1364 from PrincetonUniversity/issue-1363

Make Meshfem3D database the default database for 3D

Rohit-Kakodkar committed:
Merge pull request #1367 from PrincetonUniversity/nightly-git-info

Updated the nightly benchmarks to include git info

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1062

Rohit-Kakodkar committed:
Updates to the mesher to include external model

Rohit-Kakodkar committed:
Adds the marmousi example

icui committed:
Fix PYTHONPATH for seispie.

icui committed:
Merge pull request #1358 from PrincetonUniversity/1178-data

1178 - Update snakefile for all database in test data

icui committed:
Merge pull request #1359 from PrincetonUniversity/1178-final

1178 final - finalize displacement tests

icui committed:
Remove assemble_legacy.

icui committed:
Remove adjacency_graph::empty().

icui committed:
Update tests.

Rohit-Kakodkar committed:
Fixes GPU compilation

- [x] Adds readme instructions for marmousi

icui committed:
Merge pull request #1365 from PrincetonUniversity/1178-spin

1178 spin - Add seispie workflow for finite difference traces

icui committed:
Merge branch 'devel' into issue-1178

icui committed:
Update cmake.

icui committed:
Disable compute_tests.

Rohit-Kakodkar committed:
Issue 1379 - Deprecate Fortran generate database

- [x] Removes Fortran generate database
- [x] Removes all functions related to generate database

Rohit-Kakodkar committed:
Upgrades MESHFEM3D mesher input

- [x] Upgrades `specfem::mesh::meshfem3d` -> `specfem::mesh` namespace

Rohit-Kakodkar committed:
Documentation update - getting started & presets

Updates getting started and presets sections

Rohit-Kakodkar committed:
Documentation update - MESHFEM

Updates to mesher documentation

int-ptr-ptr committed:
added provenance to acoustic-elastic-nc

int-ptr-ptr committed:
gmshlayerbuilder in snakemake

icui committed:
Merge pull request #1406 from int-ptr-ptr/ncdisptest

Add provenance to AcousticElasticNonconformingCoupledDomain

Rohit-Kakodkar committed:
Documentation Update - Assembly, boundaries structs

- [x] Adds comprehensive documentation for assembly and boundaries struct

Rohit-Kakodkar committed:
Added code section

Rohit-Kakodkar committed:
fixed inline code

Rohit-Kakodkar committed:
Documentation Update - Assembly sources

- [x] Add documentation for `specfem::assembly::sources`

Rohit-Kakodkar committed:
Fixed intel tests

Rohit-Kakodkar committed:
Merge pull request #1377 from PrincetonUniversity/issue-1062

Issue 1062 - Fixes coupled interfaces implementation

Rohit-Kakodkar committed:
Restore previously deleted tests

Rohit-Kakodkar committed:
Added check jacobian test

lsawade committed:
Renamed simulation to program and moved context namespace.

lsawade committed:
Add new static struct specfem::MPI that holds rank, size, and comm information.

Currently it is named MPI_new, until we deprecated the old mpi struct

lsawade committed:
Forgot to add the header file

lsawade committed:
Deprecated the context - context guard combination in favor of a single context struct

lsawade committed:
Added the logger and the logger options structs

icui committed:
Merge branch 'devel' into issue-1178

icui committed:
Merge pull request #1378 from PrincetonUniversity/issue-1178

Issue 1178 - Snakemake workflow for generating test data / traces

Rohit-Kakodkar committed:
Merge pull request #1404 from PrincetonUniversity/issue-1381/2

Documentation update - getting started & presets

Rohit-Kakodkar committed:
Merge pull request #1409 from PrincetonUniversity/issue-1387-part-1

Documentation Update - Assembly, boundaries structs

Rohit-Kakodkar committed:
Update docs/sections/api/specfem/assembly/sources/locate_sources.rst

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

Rohit-Kakodkar committed:
Update docs/sections/api/specfem/assembly/sources/sort_sources_per_medium.rst

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

Rohit-Kakodkar committed:
Minor change to nc container constructor

Rohit-Kakodkar committed:
Merge pull request #1405 from PrincetonUniversity/issue-1384

Documentation update - MESHFEM

icui committed:
Merge branch 'devel' into 1178-ci

lsawade committed:
This is unfortunately doing a lot.

- This converts mpi->cout() to the appropriate specfem::Logger::<loglevel>()
- Since some of these function are being used by the tests, I needed to
  deprecate {Kokkos,MPI}_environment in favor of SPECFEM_Environement
- So alot of the tests were updated for that reason.

icui committed:
Merge pull request #1374 from PrincetonUniversity/1178-assembly

1178 - Remove assemble_legacy

Rohit-Kakodkar committed:
Merge pull request #1414 from PrincetonUniversity/issue-1401

Minor change to nc container constructor

icui committed:
Re-organize edge_tyeps.

lsawade committed:
This deprecates mpi->cout in favor of Logging methods.

The next step will be removing the specfem::MPI::MPI object and function passes.

Rohit-Kakodkar commented on core/specfem/logger/logger.cpp:
```diff
@@ -0,0 +1,316 @@
+    }
+  }
+
+  output_filename += ".txt";
```
Want to call it .log? Just a suggestion.

lsawade committed:
Merge pull request #1412 from PrincetonUniversity/issue-1387-part-2

Documentation Update - Assembly sources

icui committed:
Use graph for coupled_interface.

icui committed:
Setup coupled_interfaces in dim2.

icui committed:
Merge branch 'devel' into 1178-ci

icui committed:
Merge pull request #1403 from PrincetonUniversity/issue-1380

Upgrades MESHFEM3D mesher input

icui committed:
Update edge_types parameter.

lsawade committed:
Merge branch 'devel' into specfem-core-framework

lsawade committed:
Merge branch 'specfem-core-framework' into specfem-core-framework-depr-mpi-cout

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1379

Rohit-Kakodkar committed:
[Fix] Fixes meshfem2D for fluid solid coupling

This PR adds missing changes from [this](https://github.com/SPECFEM/specfem2d/commit/43b5f49a4ff03131961628d9a19c95d3a2cb7cea#diff-787b37144d69780313d23b0726a283f3a23cb81e4b04686afb4dff1c565997ef) commit from the original fortran code.

int-ptr-ptr committed:
gmshlayerbuilder: allow solids to skip AFS mark

int-ptr-ptr committed:
updated gmshlayerbuilder readme

Rohit-Kakodkar committed:
Merge pull request #1402 from PrincetonUniversity/issue-1379

Issue 1379 - Deprecate Fortran generate database

Rohit-Kakodkar commented on core/specfem/mpi/mpi.hpp:
```diff
@@ -0,0 +1,116 @@
+   */
+  static void check_context() {
+    if (rank == -1 || size == -1) {
+      std::cerr << "ERROR: MPI used outside Context scope" << std::endl;
+      std::exit(1);
```
We should also have an abort function within MPI. I'd think we need `abort` function within `specfem::program` that checks if we're within context before calling MPI/Kokkos abort.

Rohit-Kakodkar commented on src/specfem3d.cpp:
```diff
@@ -1,7 +1,7 @@
-#include "specfem/simulation.hpp"
-#include "specfem/simulation/context.hpp"
+#include "specfem/program.hpp"
+#include "specfem/program/context.hpp"
```
I dont think you need the `context` include. Program includes context.

Rohit-Kakodkar commented on src/specfem2d.cpp:
```diff
@@ -1,7 +1,7 @@
-#include "specfem/simulation.hpp"
-#include "specfem/simulation/context.hpp"
+#include "specfem/program.hpp"
+#include "specfem/program/context.hpp"
```
Same comment as 3d execute.

Rohit-Kakodkar commented on src/python/core.cpp:
```diff
@@ -1,5 +1,5 @@
-#include "specfem/simulation.hpp"
-#include "specfem/simulation/context.hpp"
+#include "specfem/program.hpp"
+#include "specfem/program/context.hpp"
```
Here as well.

Rohit-Kakodkar reviewed (changes requested):
We need to talk about how to safely abort code in a hybrid Kokkos/MPI code.

Rohit-Kakodkar committed:
Merge pull request #1423 from PrincetonUniversity/meshfem2d-fix

[Fix] Fixes meshfem2D for fluid solid coupling

icui committed:
Compilation fix.

icui committed:
Merge pull request #1422 from PrincetonUniversity/1178-ci

1178 - Construct assembly::coupled_interfaces from adjacency_graph

Rohit-Kakodkar committed:
Merge pull request #1426 from int-ptr-ptr/gmshlb-afsskip

gmshlayerbuilder: allow solids to skip AFS mark

lsawade committed:
Merge branch 'devel' into specfem-core-framework-depr-mpi-cout

lsawade committed:
Updated the file extension, and includes for the legacy executables

lsawade committed:
Added MPI abort functionality with line indication.

lsawade committed:
Merge branch 'specfem-core-framework-logger-impl' into specfem-core-framework-depr-mpi-cout

Rohit-Kakodkar commented on core/specfem/program/abort.cpp:
```diff
@@ -0,0 +1,58 @@
+  // Print/log error message if provided
+  if (!full_message.empty()) {
+    // Check if we're inside a valid program context
+    if (specfem::MPI_new::rank != -1 && specfem::MPI_new::size != -1) {
```
Should this be `specfem::MPI_new::check_context()`

Rohit-Kakodkar commented on core/specfem/program/abort.cpp:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  // Check if we're inside a valid program context by checking MPI_new state
+  if (specfem::MPI_new::rank != -1 && specfem::MPI_new::size != -1) {
```
Same here.

Rohit-Kakodkar reviewed (approved):
Mostly LGTM

Rohit-Kakodkar committed:
Merge pull request #1416 from PrincetonUniversity/specfem-core-framework-depr-mpi-cout

Specfem Core Framework Depr Mpi Cout

lsawade committed:
Implemented check_context for mpi abotr

lsawade committed:
Merge branch 'specfem-core-framework-logger-impl' of github.com:PrincetonUniversity/SPECFEMPP into specfem-core-framework-logger-impl
